src/scraper.rs

```diff
@@ -182,16 +183,27 @@ impl Scraper {
 
         dom.find_urls_as_strings()
             .into_iter()
-            .filter(|candidate| Scraper::should_visit(candidate, &url))
+            .filter(|candidate| Scraper::should_visit(candidate))
             .for_each(|next_url| {
-                let mut next_full_url = url.join(&next_url).unwrap();
+                let mut url_to_parse = next_url.clone();
+                // Without this line, if url is `///tools.wmflabs.org/`, the app crashes.
+                if url_to_parse.starts_with("///") { url_to_parse = url_to_parse.replacen("///", "//", 1) }
+
+                let mut next_full_url = match url.join(url_to_parse.as_str()) {
+                    Ok(url) => url,
+                    Err(e) => panic!("FAILED TO PARSE URL: {} | Error: {}", next_url, e)
+                };
+
                 next_full_url.set_fragment(None);
                 let path = url_helper::to_path(&next_full_url);
 
                 if scraper.map_url_path(&next_full_url, path.clone())
-                    && (scraper.args.depth == INFINITE_DEPTH || depth < scraper.args.depth)
                 {
-                    Scraper::push(transmitter, next_full_url, depth + 1);
+                    if !Scraper::is_on_another_domain(&next_url, &url) && (scraper.args.depth == INFINITE_DEPTH || depth < scraper.args.depth) {
+                        Scraper::push(transmitter, next_full_url, depth + 1, edepth);
+                    } else if Scraper::is_on_another_domain(&next_url, &url) && (scraper.args.edepth == INFINITE_DEPTH || edepth < scraper.args.edepth) {
+                        Scraper::push(transmitter, next_full_url, depth, edepth + 1);
```

Comment:
The condition is getting to big. Please create a sub-function to do the check

Reply:
I have instead split two ifs into four :) I am not really sure how to have those conditions as a separate functions